src/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.cs

```diff
@@ -0,0 +1,285 @@
+                    }
+                }
+
+                Counts countsBeforeUpdate = Counts.CompareExchange(ref _counts, newCounts, counts);
+                if (countsBeforeUpdate == counts)
+                {
+                    if (counts._signalCount != 0)
+                    {
+                        return true;
+                    }
+                    if (newCounts._waiterCount != counts._waiterCount)
+                    {
+                        return WaitForSignal(timeoutMs);
+                    }
+                    if (timeoutMs == 0)
+                    {
+                        return false;
+                    }
+                    break;
+                }
+
+                counts = countsBeforeUpdate;
+            }
+
+            int spinIndex = s_processorCount > 1 ? 0 : LowLevelSpinWaiter.SpinYieldThreshold;
+            while (spinIndex < LowLevelSpinWaiter.SpinCount)
+            {
+                LowLevelSpinWaiter.Wait(spinIndex);
+                spinIndex++;
+
+                counts = Counts.VolatileRead(ref _counts);
+                while (counts._signalCount > 0)
+                {
+                    Counts newCounts = counts;
+                    newCounts._signalCount--;
+                    newCounts._spinnerCount--;
+
+                    Counts countsBeforeUpdate = Counts.CompareExchange(ref _counts, newCounts, counts);
+                    if (countsBeforeUpdate == counts)
+                    {
+                        return true;
+                    }
+
+                    counts = countsBeforeUpdate;
+                }
+            }
+
+            // Unregister as spinner and acquire the semaphore or register as a waiter
+            counts = Counts.VolatileRead(ref _counts);
+            while (true)
+            {
+                Counts newCounts = counts;
+                newCounts._spinnerCount--;
+                if (counts._signalCount != 0)
+                {
+                    newCounts._signalCount--;
+                }
+                else
+                {
+                    newCounts._waiterCount++;
+                    Debug.Assert(newCounts._waiterCount != 0); // overflow check, this many waiters is currently not supported
+                }
+
+                Counts countsBeforeUpdate = Counts.CompareExchange(ref _counts, newCounts, counts);
+                if (countsBeforeUpdate == counts)
+                {
+                    return counts._signalCount != 0 || WaitForSignal(timeoutMs);
+                }
+
+                counts = countsBeforeUpdate;
+            }
+        }
+
+
+        public void Release(int releaseCount)
+        {
+            Debug.Assert(releaseCount > 0);
+
+            int countOfWaitersToWake;
+            Counts counts = Counts.VolatileRead(ref _counts);
+            while (true)
+            {
+                Counts newCounts = counts;
+
```

Comment:
Missing check for "full semaphore". It doesn't really happen now, but should be `Assert` at lease.